e2e/e2etests/test_solana_withdraw_and_call_revert.go

```diff
@@ -0,0 +1,73 @@
+package e2etests
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/gagliardetto/solana-go"
+	"github.com/stretchr/testify/require"
+	"github.com/zeta-chain/protocol-contracts/pkg/gatewayzevm.sol"
+
+	"github.com/zeta-chain/node/e2e/runner"
+	"github.com/zeta-chain/node/e2e/utils"
+	"github.com/zeta-chain/node/testutil/sample"
+	crosschaintypes "github.com/zeta-chain/node/x/crosschain/types"
+)
+
+// TestSolanaWithdrawAndCallRevert executes withdrawAndCall on zevm and calls connected program on solana
+// execution is reverted in connected program on_call function
+func TestSolanaWithdrawAndCallRevert(r *runner.E2ERunner, args []string) {
+	require.Len(r, args, 1)
+
+	withdrawAmount := utils.ParseBigInt(r, args[0])
+
+	// get ZRC20 SOL balance before withdraw
+	balanceBefore, err := r.SOLZRC20.BalanceOf(&bind.CallOpts{}, r.EVMAddress())
+	require.NoError(r, err)
+	r.Logger.Info("runner balance of SOL before withdraw: %d", balanceBefore)
+
+	require.Equal(r, 1, balanceBefore.Cmp(withdrawAmount), "Insufficient balance for withdrawal")
+
+	// parse withdraw amount (in lamports), approve amount is 1 SOL
+	approvedAmount := new(big.Int).SetUint64(solana.LAMPORTS_PER_SOL)
+	require.Equal(
+		r,
+		-1,
+		withdrawAmount.Cmp(approvedAmount),
+		"Withdrawal amount must be less than the approved amount: %v",
+		approvedAmount,
+	)
+
+	// use a random address to get the revert amount
+	revertAddress := sample.EthAddress()
+	balance, err := r.SOLZRC20.BalanceOf(&bind.CallOpts{}, revertAddress)
+	require.NoError(r, err)
+	require.EqualValues(r, int64(0), balance.Int64())
+
+	// withdraw and call
+	tx := r.WithdrawAndCallSOLZRC20(
+		runner.ConnectedProgramID,
+		withdrawAmount,
+		approvedAmount,
+		[]byte("revert"),
+		gatewayzevm.RevertOptions{
+			RevertAddress:    revertAddress,
+			OnRevertGasLimit: big.NewInt(0),
+		},
+	)
+
+	// wait for the cctx to be mined
+	cctx := utils.WaitCctxMinedByInboundHash(r.Ctx, tx.Hash().Hex(), r.CctxClient, r.Logger, r.CctxTimeout)
+	utils.RequireCCTXStatus(r, cctx, crosschaintypes.CctxStatus_Reverted)
+
+	// get ZRC20 SOL balance after withdraw
+	balanceAfter, err := r.SOLZRC20.BalanceOf(&bind.CallOpts{}, r.EVMAddress())
+	require.NoError(r, err)
+	r.Logger.Info("runner balance of SOL after withdraw: %d", balanceAfter)
+
+	// check the balance of revert address is equal to withdraw amount
+	balance, err = r.SOLZRC20.BalanceOf(&bind.CallOpts{}, revertAddress)
+	require.NoError(r, err)
+
+	require.True(r, balance.Cmp(withdrawAmount) == 0)
```

Comment:
Nit: better to convert to uint64 and check the values so you directly see the different if the test fails